File: bfs-web-app/types/product.ts
import type { CategoryDTO } from './category';
import type { Cents } from './common';
import type { MenuDTO } from './menu';

export type ProductType = "simple" | "menu";

export interface Product {
  id: string;
  categoryId: string;
  type: ProductType;
  name: string;
  image: string | null;
  priceCents: Cents;
  isActive: boolean;
  createdAt: string; // ISO date
  updatedAt: string; // ISO date
}

export interface ProductSummaryDTO {
  id: string;
  category: CategoryDTO;
  type: ProductType;
  name: string;
  image: string | null;
  priceCents: Cents;
  isActive: boolean;
  availableQuantity?: number | null;
  isAvailable?: boolean;
  isLowStock?: boolean;
}

export interface ProductDTO extends ProductSummaryDTO {
  menu: MenuDTO | null;
}

File: bfs-backend/internal/domain/product.go
package domain

import (
	"time"

	"go.mongodb.org/mongo-driver/v2/bson"
)

type ProductType string

const (
	ProductTypeSimple ProductType = "simple"
	ProductTypeMenu   ProductType = "menu"
)

type Product struct {
	ID         bson.ObjectID `bson:"_id"`
	CategoryID bson.ObjectID `bson:"category_id" validate:"required"`
	Type       ProductType   `bson:"type" validate:"required,oneof=simple menu"`
	Name       string        `bson:"name" validate:"required"`
	Image      *string       `bson:"image,omitempty"`
	PriceCents Cents         `bson:"price_cents" validate:"required,gte=0"`
	IsActive   bool          `bson:"is_active"`
	CreatedAt  time.Time     `bson:"created_at"`
	UpdatedAt  time.Time     `bson:"updated_at"`
}

type ProductSummaryDTO struct {
	ID         string      `json:"id"`
	Category   CategoryDTO `json:"category"`
	Type       ProductType `json:"type"`
	Name       string      `json:"name"`
	Image      *string     `json:"image,omitempty"`
	PriceCents Cents       `json:"priceCents"`
	IsActive   bool        `json:"isActive"`
	// Inventory snapshot for availability on menus
	AvailableQuantity *int  `json:"availableQuantity,omitempty"`
	IsAvailable       *bool `json:"isAvailable,omitempty"`
	IsLowStock        *bool `json:"isLowStock,omitempty"`
}

type ProductDTO struct {
	ProductSummaryDTO
	Menu *MenuDTO `json:"menu,omitempty"`
}
